docs/proposals/synchronous-garbage-collection.md

```diff
@@ -0,0 +1,126 @@
+
+**kubelet** deletes the pod from the key-value store after all its containers are terminated ([code](../../pkg/kubelet/status/status_manager.go#L441-L443)). It also assumes that if the API server does not return an error, the pod is removed from the key-value store. Breaking the assumption will not break `kubelet` though, because the `pod` must have already been in the terminated `phase`, `kubelet` will not care to manage it.
+
+**Node controller** forcefully deletes pod if the pod is scheduled to a node that does not exist ([code](../../pkg/controller/node/nodecontroller.go#L474)). The pod will continue to exist if it has pending finalizers. The node controller will futilely retry the deletion. Also, the `node controller` forcefully deletes pods before deleting the node ([code](../../pkg/controller/node/nodecontroller.go#L592)). If the pods have pending finalizers, the `node controller` will go ahead deleting the node, leaving those pods behind. These pods will be deleted from the key-value store when the pending finalizers are removed.
+
+**Podgc** deletes terminated pods if there are too many of them in the cluster. `Podgc` should remove any pending finalizers to make sure the pods are deleted.
+
+**Deployment controller** adopts existing `ReplicaSet` (RS) if its template matches. If a matching RS has a pending `GCFinalizer`, deployment shouldn't adopt it, because the RS controller will not scale up/down a RS that's being deleted. Hence, `deployment controller` needs to check if a RS is being deleted before adopting it. If the RS is being deleted, then the `deployment controller` should wait for the status of the RS to show 0 replicas to avoid creating extra pods, then create a new RS.
+
+**Replication controller manager**, **Job controller**, and **ReplicaSet controller** ignore pods in terminated phase, so pods with pending finalizers will not block these controllers.
+
+**PetSet controller** will be blocked by a pod with pending finalizers, so Synchronous GC might slow down its progress.
+
+**kubectl**: synchronous GC can replace the **kubectl delete** reapers. Currently `kubectl delete` blocks until all dependents and the owner are deleted. To maintain this behavior, after switched to using synchronous GC, *kubectl delete* needs to poll on the removal of the owner object.
+
+## Security implications
+
+A user who is authorized to update one object can affect the synchronous GC behavior of another object. Specifically, by setting an object as a pod's owner, and setting a very long grace termination period for the pod, a user can make the synchronous GC of the owner to take long time.
```

Comment:
I think there's a number of ways to reduce the severity of this problem. These are just some ideas.

1. apiserver could control the list of finalizers you're allowed to add to an object. You'd have to register a finalizer and only sysadmins would have permission to do that. This reduces the chance of an attack where you keep an object around by adding a bogus finalizer.
2. If we go with the finalizer registry idea, you could supply a max time along with the finalizer (probably bad idea for a number of reasons).
3. We could make adding a finalizer require a particular permission, although this would be a field level auth. We might need a dedicated subresource for this.
4. We could consider adding a restriction, like "cross-user links are {not permitted by apiserver | not honored by sync deletion by the GC}"
5. We could add a restriction like, you can't add a link to an object that can't see you. Although this is pretty weird because it means less permissions is better for the attackee.
6. Maybe better than that, just require mutual visibility.
7. Actually, even better-- make the sync GC mode do a permission check. If Ob->Oa, and Oa is sync deleted, but Oa cannot see Ob, then still go ahead and delete Ob, but don't consider it as something that needs to be waited for. This still lets people in some namespace get their thing turned down when a dependency is turned down, but they can't block the dependency's turndown process.
8. We could surface a metric, time spent in sync GC. sysadmins could get paged if it starts going up. (we should surface this metric no matter what, arguably.)